src/reanimated2/helperTypes.ts

```diff
@@ -8,74 +7,58 @@ This will not be easy though!
 */
 
 import type {
-  ColorValue,
+  ImageStyle,
+  RegisteredStyle,
   StyleProp,
-  TransformsStyle as RNTransformsStyle,
+  TextStyle,
+  TransformsStyle,
+  ViewStyle,
 } from 'react-native';
+import type { AnimatableValue, SharedValue } from './commonTypes';
+import type { BaseAnimationBuilder } from './layoutReanimation/animationBuilder/BaseAnimationBuilder';
 import type {
-  AnimatableValue,
-  BaseAnimationBuilder,
   EntryExitAnimationFunction,
   LayoutAnimationFunction,
-  SharedValue,
-} from '.';
+} from './layoutReanimation/animationBuilder/commonTypes';
 import type { ReanimatedKeyframe } from './layoutReanimation/animationBuilder/Keyframe';
 import type { SharedTransition } from './layoutReanimation/sharedTransitions';
 import type { DependencyList } from './hook/commonTypes';
 
-export type Adaptable<T> =
-  | T
-  | ReadonlyArray<T | ReadonlyArray<T>>
-  | SharedValue<T>;
+export type TransformArrayItem = NonNullable<
+  TransformsStyle['transform']
+>[number];
 
-export type AdaptTransforms<T> = {
-  [P in keyof T]: Adaptable<T[P]>;
-};
-
-type TransformsStyle = Pick<RNTransformsStyle, 'transform'>;
+export type AnimatedTransform = MaybeSharedValueRecursive<
+  TransformsStyle['transform']
+>;
 
-export type TransformStyleTypes = TransformsStyle['transform'] extends
-  | readonly (infer T)[]
-  | string
-  | undefined
-  ? T
+type MaybeSharedValue<Value> = Value | Value extends AnimatableValue
+  ? SharedValue<Value>
   : never;
-export type AnimatedTransform = AdaptTransforms<TransformStyleTypes>[];
 
-/**
- * @deprecated Please use `AnimatedStyle` type instead.
- */
-export type AnimateStyle<S> = {
-  [K in keyof S]: K extends 'transform'
-    ? AnimatedTransform
-    : S[K] extends ReadonlyArray<any>
-    ? ReadonlyArray<AnimateStyle<S[K][0]>>
-    : S[K] extends object
-    ? AnimateStyle<S[K]>
-    : S[K] extends ColorValue | undefined
-    ? S[K] | number
-    : S[K] | SharedValue<AnimatableValue>;
-};
+type MaybeSharedValueRecursive<Value> = Value extends (infer InnerValue)[]
```

Comment:
It would be more readable if we show that we are trying to obtain an array element
```suggestion
type MaybeSharedValueRecursive<Value> = Value extends Array<infer ArrayElement>
```

Reply:
That's just a syntax preference here and honestly I'm leaning more towards the `[]` notation.